portallocator/portallocator.go

```diff
@@ -165,6 +165,14 @@ func (p *PortAllocator) ReleasePort(ip net.IP, proto string, port int) error {
 	if !ok {
 		return nil
 	}
+	for _, pr := range protomap[proto].portRanges {
```

Comment:
Instead of this loop, I was suggesting something very very simple.
if port < pr.last {
  pr.last = port
}
That should be sufficient to reallocate the smallest, and on port allocation, when last is allocated, last will be last++ (you have to see the corner cases at the border of the interval, maybe you have to keep last within the boundaries)